### PR TITLE
[Enhancement](Compaction) Caculate all committed rowsets delete bitmaps when do comapction

### PR DESCRIPTION
**Pull request by Yukang-Lian**
## Proposed changes

Issue Number: close #xxx

Here we will calculate all the rowsets delete bitmaps which are committed but not published to reduce the calculation pressure of publish phase.

Step1: collect this tablet's all committed rowsets' delete bitmaps.

Step2: calculate all rowsets' delete bitmaps which are published during compaction.

Step3: write back updated delete bitmap and tablet info.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by Yukang-Lian** [comment]
run p0

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 37.61 seconds
 stream load tsv:          457 seconds loaded 74807831229 Bytes, about 156 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          63.9 seconds inserted 10000000 Rows, about 156K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230616112909_clickbench_pr_163691.html

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

File: be/src/olap/compaction.cpp [hunk 1]
```diff
@@ -529,8 +532,58 @@ Status Compaction::modify_rowsets(const Merger::Statistics* stats) {
...
+                RETURN_IF_ERROR(beta_rowset->load_segments(&segments));
+                RETURN_IF_ERROR(_tablet->commit_phase_update_delete_bitmap(
+                        tablet_txn_info.rowset, tablet_txn_info.rowset_ids,
+                        tablet_txn_info.delete_bitmap, cur_max_version, segments, it.first.second,
+                        delta_writer->get_rowset_writer()));
+                // Step3: write back updated delete bitmap and tablet info.
+                StorageEngine::instance()->txn_manager()->set_txn_related_delete_bitmap(
+                        it.first.first, it.first.second, _tablet->tablet_id(),
+                        _tablet->schema_hash(), _tablet->tablet_uid(), true,
+                        tablet_txn_info.delete_bitmap, _tablet->all_rs_id(cur_max_version));
```
Comment (zhannngchen) [inline be/src/olap/compaction.cpp:579]: 1. you should update the converted delete bitmap to txn_info
2. add the compacted rowset's id to txn_info.rowset_ids

File: be/src/olap/compaction.cpp [hunk 2]
```diff
@@ -529,8 +532,58 @@ Status Compaction::modify_rowsets(const Merger::Statistics* stats) {
             }
         }
 
+        // Here we will calculate all the rowsets delete bitmaps which are committed but not published to reduce the calculation pressure
+        // of publish phase.
+        // All rowsets which need to recalculate have been published so we don't need to acquire lock.
+        // Step1: collect this tablet's all committed rowsets' delete bitmaps
+        TxnManager::txn_tablet_map_t txn_tablet_map {};
```
Comment (zhannngchen) [inline be/src/olap/compaction.cpp:539]: we'd better to move this part after L589, do these works with lock, which can avoid potential consistency issues

File: be/src/olap/compaction.cpp [hunk 3]
```diff
@@ -540,7 +593,7 @@ Status Compaction::modify_rowsets(const Merger::Statistics* stats) {
             // incremental data.
             _tablet->calc_compaction_output_rowset_delete_bitmap(
                     _input_rowsets, _rowid_conversion, version.second, UINT64_MAX, &missed_rows,
-                    &location_map, &output_rowset_delete_bitmap);
+                    &location_map, &output_rowset_delete_bitmap, commit_rowset_delete_bitmap);
```
Comment (zhannngchen) [inline be/src/olap/compaction.cpp:596]: you should not merge these un-published bitmap to Tablet's bitmap

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

File: be/src/olap/tablet.cpp [hunk 4]
```diff
@@ -3296,29 +3295,39 @@ void Tablet::calc_compaction_output_rowset_delete_bitmap(
...
-                    output_rowset_delete_bitmap->add({dst.rowset_id, dst.segment_id, cur_version},
-                                                     dst.row_id);
-                }
+            convert_rowid(rowset, subset_map, src, rowid_conversion, start_version, end_version,
+                          missed_rows, location_map, output_rowset_delete_bitmap);
+        }
+    }
+}
+
+void Tablet::convert_rowid(
```
Comment (zhannngchen) [inline be/src/olap/tablet.cpp:3304]: This method don't needs so many parameters, we should try our best to make the declaration more clear

File: be/src/olap/txn_manager.cpp [hunk 5]
```diff
@@ -653,6 +653,19 @@ void TxnManager::get_all_related_tablets(std::set<TabletInfo>* tablet_infos) {
     }
 }
 
+void TxnManager::get_all_commit_tablet_txn_info_by_tablet(const TabletSharedPtr& tablet,
+                                                          txn_tablet_map_t& txn_tablet_map) {
```
Comment (zhannngchen) [inline be/src/olap/txn_manager.cpp:657]: seems we just need a list, using a map here is confusing

File: be/src/olap/compaction.cpp [hunk 6]
```diff
@@ -574,11 +575,47 @@ Status Compaction::modify_rowsets(const Merger::Statistics* stats) {
...
+            for (const auto& it : txn_tablet_map) {
+                for (const auto& tablet_load_it : it.second) {
+                    const TabletTxnInfo& tablet_txn_info = tablet_load_it.second;
+                    DeleteBitmap output_delete_bitmap(_tablet->tablet_id());
+                    RowLocation src;
+                    std::shared_ptr<Rowset> rowset = tablet_txn_info.rowset;
+                    src.rowset_id = rowset->rowset_id();
+                    for (uint32_t seg_id = 0; seg_id < rowset->num_segments(); ++seg_id) {
+                        src.segment_id = seg_id;
+                        _tablet->convert_rowid(rowset, *tablet_txn_info.delete_bitmap, src,
```
Comment (zhannngchen) [inline be/src/olap/compaction.cpp:605]: can't use such delete bitmap directly

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run p0

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run clickbench

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 43.88 seconds
 stream load tsv:          459 seconds loaded 74807831229 Bytes, about 155 MB/s
 stream load json:         24 seconds loaded 2358488459 Bytes, about 93 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 insert into select:          70.7 seconds inserted 10000000 Rows, about 141K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230629150723_clickbench_pr_170061.html

File: be/src/olap/compaction.cpp [hunk 7]
```diff
@@ -574,16 +577,47 @@ Status Compaction::modify_rowsets(const Merger::Statistics* stats) {
...
+            // Step2: calculate all rowsets' delete bitmaps which are published during compaction.
+            int64_t cur_max_version = _tablet->max_version().second;
+            RowsetIdUnorderedSet rowset_ids = _tablet->all_rs_id(cur_max_version);
+            rowset_ids.insert(_output_rowset->rowset_id());
+            for (const auto& it : commit_tablet_txn_info_vec) {
+                DeleteBitmap output_delete_bitmap(_tablet->tablet_id());
+                const std::shared_ptr<Rowset>& rowset = it.rowset;
+                for (uint32_t seg_id = 0; seg_id < rowset->num_segments(); ++seg_id) {
+                    _tablet->calc_compaction_output_rowset_delete_bitmap(
+                            std::vector<std::shared_ptr<Rowset>>(1, rowset), _rowid_conversion,
```
Comment (zhannngchen) [inline be/src/olap/compaction.cpp:603]: You used wrong input rowset?

File: be/src/olap/compaction.cpp [hunk 8]
```diff
@@ -574,16 +577,47 @@ Status Compaction::modify_rowsets(const Merger::Statistics* stats) {
...
+            // of publish phase.
+            // All rowsets which need to recalculate have been published so we don't need to acquire lock.
+            // Step1: collect this tablet's all committed rowsets' delete bitmaps
+            CommitTabletTxnInfoVec commit_tablet_txn_info_vec {};
+            StorageEngine::instance()->txn_manager()->get_all_commit_tablet_txn_info_by_tablet(
+                    _tablet, commit_tablet_txn_info_vec);
+
+            // Step2: calculate all rowsets' delete bitmaps which are published during compaction.
+            int64_t cur_max_version = _tablet->max_version().second;
+            RowsetIdUnorderedSet rowset_ids = _tablet->all_rs_id(cur_max_version);
```
Comment (zhannngchen) [inline be/src/olap/compaction.cpp:596]: should not replace the rowset_ids with `_tablet->all_rs_id(cur_max_version)`

**Comment by Yukang-Lian** [comment]
run buildall

File: be/src/olap/txn_manager.cpp [hunk 9]
```diff
@@ -632,6 +632,28 @@ void TxnManager::get_all_related_tablets(std::set<TabletInfo>* tablet_infos) {
     }
 }
 
+void TxnManager::get_all_commit_tablet_txn_info_by_tablet(
```
Comment (liaoxin01) [inline be/src/olap/txn_manager.cpp:635]: Just returning references to delete_bitmap and rowset_ids is enough?

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 50.77 seconds
 stream load tsv:          453 seconds loaded 74807831229 Bytes, about 157 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          70.3 seconds inserted 10000000 Rows, about 142K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230706165641_clickbench_pr_173946.html

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 52.16 seconds
 stream load tsv:          511 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          87.1 seconds inserted 10000000 Rows, about 114K ops/s
 storage size: 17164874736 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707012436_clickbench_pr_173942.html

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.85 seconds
 stream load tsv:          455 seconds loaded 74807831229 Bytes, about 156 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          58 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          69.1 seconds inserted 10000000 Rows, about 144K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707024729_clickbench_pr_174066.html

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 54.63 seconds
 stream load tsv:          512 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          68 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          85.3 seconds inserted 10000000 Rows, about 117K ops/s
 storage size: 17162237060 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707105824_clickbench_pr_174064.html

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 50.85 seconds
 stream load tsv:          507 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          87.9 seconds inserted 10000000 Rows, about 113K ops/s
 storage size: 17162459196 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707163542_clickbench_pr_174340.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 54.79 seconds
 stream load tsv:          457 seconds loaded 74807831229 Bytes, about 156 MB/s
 stream load json:         23 seconds loaded 2358488459 Bytes, about 97 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 insert into select:          67.9 seconds inserted 10000000 Rows, about 147K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707083938_clickbench_pr_174349.html

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 56.05 seconds
 stream load tsv:          507 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          89.2 seconds inserted 10000000 Rows, about 112K ops/s
 storage size: 17167907865 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707184029_clickbench_pr_174492.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 51.85 seconds
 stream load tsv:          457 seconds loaded 74807831229 Bytes, about 156 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 insert into select:          71.0 seconds inserted 10000000 Rows, about 140K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707113652_clickbench_pr_174489.html

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 49.55 seconds
 stream load tsv:          454 seconds loaded 74807831229 Bytes, about 157 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 insert into select:          68.8 seconds inserted 10000000 Rows, about 145K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230708040153_clickbench_pr_174788.html

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 55.75 seconds
 stream load tsv:          508 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          66 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          88.5 seconds inserted 10000000 Rows, about 112K ops/s
 storage size: 17162241855 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230708121445_clickbench_pr_174790.html

**Review by zhannngchen — APPROVED** [review]
LGTM

**Review by liaoxin01 — APPROVED** [review]
LGTM

**Pull request merged**